rama-net/src/fingerprint/peet/tls.rs

```diff
@@ -0,0 +1,375 @@
+        let mut psk_key_exchange_mode = None;
+        let mut certificate_compression_algorithms = Vec::new();
+        let mut sorted_extensions = Vec::new();
+
+        for ext in client_hello.extensions() {
+            let id = ext.id();
+
+            if id.is_grease() {
+                sorted_extensions.push("GREASE".to_owned());
+                continue;
+            }
+
+            sorted_extensions.push(u16::from(id).to_string());
+
+            match ext {
+                ClientHelloExtension::SupportedVersions(versions) => {
+                    supported_tls_versions.extend(versions.iter().copied());
+                }
+                ClientHelloExtension::ApplicationLayerProtocolNegotiation(alpns) => {
+                    for p in alpns {
+                        match p {
+                            ApplicationProtocol::HTTP_10 => {}
+                            ApplicationProtocol::HTTP_11 => {}
+                            ApplicationProtocol::HTTP_2 => {}
+                            _ => {
+                                continue;
+                            }
+                        }
+
+                        supported_protocols.push(p.clone());
+                    }
+                }
+                ClientHelloExtension::SupportedGroups(groups) => {
+                    supported_groups.extend(groups.iter().copied());
+                }
+                ClientHelloExtension::SignatureAlgorithms(algos) => {
+                    supported_signature_algorithms.extend(algos.iter().copied());
+                }
+                ClientHelloExtension::Opaque { id: ext_id, data }
+                    if *ext_id == ExtensionId::from(45) =>
+                {
+                    if data.len() < 2 {
+                        psk_key_exchange_mode = None;
+                    } else {
+                        psk_key_exchange_mode = Some(data[1]);
+                    }
+                }
+                ClientHelloExtension::CertificateCompression(algs) => {
+                    certificate_compression_algorithms.extend(algs.iter().cloned());
+                }
+                _ => {}
+            }
+        }
+
+        sorted_extensions.sort();
+
+        Ok(Self {
+            supported_tls_versions,
+            supported_protocols,
+            supported_groups,
+            supported_signature_algorithms,
+            psk_key_exchange_mode,
+            certificate_compression_algorithms: if certificate_compression_algorithms.is_empty() {
+                None
+            } else {
+                Some(certificate_compression_algorithms)
+            },
+            cipher_suites,
+            sorted_extensions,
+        })
+    }
+
+    #[inline]
+    pub fn to_human_string(&self) -> String {
+        format!("{self:?}")
+    }
+
+    pub fn fmt_as(&self, f: &mut fmt::Formatter<'_>, hash_chunks: bool) -> fmt::Result {
+        let tls_versions_str: String = self
```

Comment:
try to avoid allocating so much, by just writing to it already. So iterate over tls versions and use stuff like `write!(f, "GREASE")`, `write!(f, "{v}")` etc ... Also less collecting into vecs and joining etc... Just write the separators :)

Reply:
And once you go for this write on the fly approach instead of intermediate allocations you would simply wrap the formatter with an md5 encoder :) Your underlying code in the end just cares about it being an something that you can use `write!` with. You can check ja3.rs to see how you can use a macro to make it work for both a hasher and a formatter, as these are sadly not compattible afaik, unless you find a cleaner way.